R/config-log.R

```diff
@@ -168,18 +170,58 @@ config_log_entry <- function(path,
     warning(msg)
     return(NULL)
   }
+  config[["nm_version"]] <- resolve_nonmem_version(config) %||% NA_character_
 
-  matches <- suppressMessages(check_up_to_date(cfg_mod))
+  return(list(config = config, fields = c(fields, "nm_version")))
+}
+
+#' Parse a bbi config file
+#'
+#' @param path A string giving the path to `bbi_config.json`.
+#' @param fields A character vector of fields to include.
+#'
+#' @return A list whose elements include
+#'
+#'   * the path to the model file (minus extension)
+#'
+#'   * `fields`
+#'
+#'   * whether the model file has changed
+#'
+#'   * whether the data file has changed
+#'
+#'   * the version of NONMEM
+#'
+#'   The return value is `NULL` if any element of `fields` is not found in
+#'   `path`.
+#'
+#' @keywords internal
+config_log_entry <- function(path, fields = NULL) {
+  checkmate::assert_file_exists(path)
+  checkmate::assert_character(fields, null.ok = TRUE)
+
+  cfg_mod <- read_model_from_config(path)
+  config <- jsonlite::fromJSON(path)
+
+  res <- config_log_make_entry(cfg_mod, config, fields)
+  if (is.null(res$config)) {
+    return(NULL)
+  }
+
+  config <- res$config
+  fields <- res$fields
+
+  # bbr.bayes kludge: .build_data is passed for
+  # check_up_to_date.bbi_stan_{model,summary}.
+  matches <- suppressMessages(check_up_to_date(cfg_mod, .build_data = TRUE))
```

Comment:
Hmm, so that line comes from before the bbr.bayes split off: 1ade325f (feat: made config_log() work with bbi_stan_model objects. Sets .build_data = TRUE for Stan models. "...", 2021-03-18).  In f065f2af (add comments to point out two bbr.bayes related kludges, 2022-08-12), I added the comment and said it was "probably not worth worrying about getting rid of at this point".

So, I wish I would have spelled at my rationale more, but looking at it now I'd say it is the combination 1) first making the minimal change from the pre-split state to make it work and then 2) deciding to leave it as is because the leakage isn't too bad and it avoids some amount of duplication across `config_log_make_entry` methods.

But I'm not against leaving it to `config_log_make_entry` methods to handle.  That'd involve updating the methods (including bbr.bayes), documenting this expectation in the `config_log_make_entry` generic, and adding some guards to `config_log_entry` to make sure the method has added the "*_has_changed" fields.

Would you like to go that route?

Reply:
Ok, that makes sense. Let's leave it as is. Some thoughts/explanations below.

### thought 1

> _from me:_ with the irrelevant .build_data arg

I now see that's wrong. It is relevant, because we're just doing the more strict version of the check (where you actually rebuild the data). And this arg just falls through the dots for NONMEM (and other models that it isn't relevant for) so that's fine.

### thought 2

> so that line comes from before the bbr.bayes split off...

My fault, I misread this diff:

<img width="858" alt="Screenshot 2023-03-15 at 11 10 15 AM" src="https://user-images.githubusercontent.com/7806083/225353632-f83c9f89-f37d-4ccf-bcf1-a6c3bde229b3.png">

I thought that you had intentionally moved it out of there.

### thought 3

> [current state] avoids some amount of duplication across `config_log_make_entry` methods... But I'm not against leaving it to `config_log_make_entry` methods to handle. That'd involve updating the methods (including `bbr.bayes`)...

This is fair. I hadn't considered that we would need to be sure to do this in every `config_log_make_entry` method. As it is now, it should just work for all methods, so let's leave it as is for now. If ever stumble onto a case where it should be handled differently for some class of models, then we could potentially revisit. (Probably unlikely.)